testbed/tests/window/test_window.py

```diff
@@ -793,24 +791,30 @@ async def test_window_state_rapid_assignment(
         second_window.toolbar.add(app.cmd1)
         second_window.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
         second_window.show()
-        # Add delay to ensure windows are visible after animation.
+        # Wait for window animation before assertion.
         await second_window_probe.wait_for_window("Secondary window is visible")
         assert second_window_probe.instantaneous_state == WindowState.NORMAL
 
         # Assign new states without waiting for each transition to complete,
         # to test that the backend can handle rapid window state assignments.
         for state in states:
             second_window.state = state
-
-        # Add delay to ensure windows are visible after the final state animation.
+        # Wait for window animation before assertion.
         await second_window_probe.wait_for_window(
-            f"Secondary window is in {states[-1]}", state_switch_not_from_normal=True
+            f"Secondary window is in {states[-1]}", expected_state=states[-1]
         )
 
         # Verify that the backend handled rapid assignments by checking if
         # the window reached the correct final window state.
         assert second_window_probe.instantaneous_state == states[-1]
 
+        # Required to ensure complete coverage of fullscreen handling on macOS.
+        if states[-1] == WindowState.FULLSCREEN:
+            await second_window_probe.wait_for_window(
+                "Waiting for completion of fullscreen pending state handling",
+                full_screen=True,
+            )
+
```

Comment:
This is required to ensure coverage on line 99:
https://github.com/beeware/toga/blob/c69f5e4910155b2ef9655a9d2c5ef81d6ae84df1/cocoa/src/toga_cocoa/window.py#L88-L100

This is because, wait_for_window() waits only for the window to switch to FULLSCREEN(i.e., states[-1]), and then the test is completed. Then window is closed in window_cleanup. wait_for_window() doesn't wait for windowDidEnterFullScreen_ notification, and so line 99 doesn't get the chance to execute. 

Earlier, line 99 was getting covered as we were using fixed length delays instead of polling, and so the test waited long enough for windowDidEnterFullScreen_ to be notified to the window and allowed line 99 to be executed.

Reply:
I understand the reasoning, but I don't like the approach you've taken here. AFAICT, the issue is that the assertion on window states is evaluating the *instantaneous* state, and not checking for pending state.

Reply:
I agree that's the issue. Could wait_for_window() include an optional assertion_test_method() parameter? This would allow checking additional window properties beyond just the state.

Reply:
But, I think since the condition to ensure coverage is much simpler and only a single assertion, so I have ensured coverage by doing the assertion at the window probe.